File: Backend/app/services/interviews.py
from app.core.ids import IdPrefix, new_id
from app.errors.codes import ErrorCode
from app.errors.exceptions import NotFoundError
from app.repositories.interviews import InterviewRepository
from app.schemas.common import InterviewStatus, RoundStatus
from app.schemas.interviews import CreateInterviewRequest, Interview, UpdateInterviewRequest

_NOT_FOUND_MESSAGE = "That interview could not be found."


class InterviewService:
    def __init__(self, interviews: InterviewRepository) -> None:
        self._interviews = interviews

    async def list_for_user(self, user_id: str) -> list[Interview]:
        docs = await self._interviews.list_for_user(user_id)
        return [Interview(**doc) for doc in docs]

    async def get(self, user_id: str, interview_id: str) -> Interview:
        doc = await self._interviews.get(user_id, interview_id)
        if doc is None:
            raise NotFoundError(ErrorCode.INTERVIEW_NOT_FOUND, _NOT_FOUND_MESSAGE)
        return Interview(**doc)

    async def create(self, user_id: str, request: CreateInterviewRequest) -> Interview:
        doc = {
            "id": new_id(IdPrefix.INTERVIEW),
            "user_id": user_id,
            "company": request.company,
            "company_mark": request.company[:1].upper(),
            "role": request.role,
            "type": request.type,
            "round": "Current round",
            "round_number": 1,
            "total_rounds": 1,
            "scheduled_at": request.scheduled_at,
            "timezone": request.timezone,
            "duration_minutes": 60,
            "meeting_url": None,
            "recruiter": None,
            "interviewers": None,
            "status": InterviewStatus.UPCOMING,
            "readiness": 0,
            "preparation_progress": 0,
            "location": "Not added",
            "accent": "#f0b94c",
            "rounds": [
                {
                    "id": new_id(IdPrefix.ROUND),
                    "name": "Current round",
                    "type": request.type,
                    "status": RoundStatus.CURRENT,
                }
            ],
            "provider_event_id": None,
        }
        await self._interviews.insert(doc)
        return Interview(**doc)

    async def update(
        self, user_id: str, interview_id: str, request: UpdateInterviewRequest
    ) -> Interview:
        changes = request.model_dump(exclude_unset=True)
        doc = await self._interviews.update(user_id, interview_id, changes)
        if doc is None:
            raise NotFoundError(ErrorCode.INTERVIEW_NOT_FOUND, _NOT_FOUND_MESSAGE)
        return Interview(**doc)

    async def confirm(self, user_id: str, interview_id: str) -> Interview:
        doc = await self._interviews.update(
            user_id, interview_id, {"status": InterviewStatus.CONFIRMED}
        )
        if doc is None:
            raise NotFoundError(ErrorCode.INTERVIEW_NOT_FOUND, _NOT_FOUND_MESSAGE)
        return Interview(**doc)

    async def delete(self, user_id: str, interview_id: str) -> None:
        deleted = await self._interviews.delete(user_id, interview_id)
        if not deleted:
            raise NotFoundError(ErrorCode.INTERVIEW_NOT_FOUND, _NOT_FOUND_MESSAGE)

File: src/mocks/handlers/interviews.ts
import { http, HttpResponse } from "msw";

import { buildDashboardOverview } from "@/mocks/fixtures";
import { db, nextId } from "@/mocks/db";
import type { Interview, InterviewRound } from "@/types/domain";

function notFound(requestId?: string) {
  return HttpResponse.json(
    {
      error: {
        code: "INTERVIEW_NOT_FOUND",
        message: "That interview could not be found.",
        details: {},
        requestId,
      },
    },
    { status: 404 },
  );
}

/** See docs/API-CONTRACT.md's Interviews section. */
export const interviewHandlers = [
  http.get("*/dashboard/overview", () => HttpResponse.json(buildDashboardOverview(db.interviews))),

  http.get("*/interviews", () => HttpResponse.json(db.interviews)),

  http.get("*/interviews/:id", ({ params }) => {
    const interview = db.interviews.find((item) => item.id === params.id);
    return interview ? HttpResponse.json(interview) : notFound();
  }),

  http.post("*/interviews", async ({ request }) => {
    const body = (await request.json()) as Pick<
      Interview,
      "company" | "role" | "type" | "scheduledAt" | "timezone"
    >;
    const round: InterviewRound = {
      id: nextId("round"),
      name: "Current round",
      type: body.type,
      status: "current",
    };
    const interview: Interview = {
      id: nextId("interview"),
      company: body.company,
      companyMark: body.company.slice(0, 1).toUpperCase(),
      role: body.role,
      type: body.type,
      round: round.name,
      roundNumber: 1,
      totalRounds: 1,
      scheduledAt: body.scheduledAt,
      timezone: body.timezone,
      durationMinutes: 60,
      status: "upcoming",
      readiness: 0,
      preparationProgress: 0,
      location: "Not added",
      accent: "#f0b94c",
      rounds: [round],
    };
    db.interviews = [interview, ...db.interviews];
    return HttpResponse.json(interview, { status: 201 });
  }),

  http.patch("*/interviews/:id", async ({ params, request }) => {
    const index = db.interviews.findIndex((item) => item.id === params.id);
    const current = db.interviews[index];
    if (!current) return notFound();
    const body = (await request.json()) as Partial<Interview>;
    const updated: Interview = { ...current, ...body, id: current.id };
    db.interviews = db.interviews.map((item, itemIndex) => (itemIndex === index ? updated : item));
    return HttpResponse.json(updated);
  }),

  http.delete("*/interviews/:id", ({ params }) => {
    const exists = db.interviews.some((item) => item.id === params.id);
    if (!exists) return notFound();
    db.interviews = db.interviews.filter((item) => item.id !== params.id);
    return new HttpResponse(null, { status: 204 });
  }),

  http.post("*/interviews/:id/confirm", ({ params }) => {
    const index = db.interviews.findIndex((item) => item.id === params.id);
    if (index === -1) return notFound();
    const updated: Interview = { ...db.interviews[index]!, status: "confirmed" };
    db.interviews = db.interviews.map((item, itemIndex) => (itemIndex === index ? updated : item));
    return HttpResponse.json(updated);
  }),
];
